File: src/java-sample/com/xtaudio/xt/PrintDetailed.java
package com.xtaudio.xt;

public class PrintDetailed {

    static void onFatal() {
        System.out.println("Fatal error.");
    }

    static void onTrace(XtLevel level, String message) {
        if (level != XtLevel.INFO)
            System.out.println("-- " + level + ": " + message);
    }

    public static void main(String[] args) {

        try (XtAudio audio = new XtAudio("Sample", null, PrintDetailed::onTrace, PrintDetailed::onFatal)) {

            System.out.println("Win32: " + XtAudio.isWin32());
            System.out.println("Version: " + XtAudio.getVersion());
            System.out.println("Pro Audio: " + XtAudio.getServiceBySetup(XtSetup.PRO_AUDIO).getName());
            System.out.println("System Audio: " + XtAudio.getServiceBySetup(XtSetup.SYSTEM_AUDIO).getName());
            System.out.println("Consumer Audio: " + XtAudio.getServiceBySetup(XtSetup.CONSUMER_AUDIO).getName());

            for (int s = 0; s < XtAudio.getServiceCount(); s++) {

                XtService service = XtAudio.getServiceByIndex(s);
                System.out.println("Service " + service.getName() + ":");
                System.out.println("  System: " + service.getSystem());
                System.out.println("  Device count: " + service.getDeviceCount());
                System.out.println("  Capabilities: " + XtPrint.capabilitiesToString(service.getCapabilities()));

                try (XtDevice defaultInput = service.openDefaultDevice(false)) {
                    System.out.println("  Default input: " + defaultInput);
                }
                try (XtDevice defaultOutput = service.openDefaultDevice(true)) {
                    System.out.println("  Default output: " + defaultOutput);
                }

                for (int d = 0; d < service.getDeviceCount(); d++)
                    try (XtDevice device = service.openDevice(d)) {

                        System.out.println("  Device " + device.getName() + ":");
                        System.out.println("    System: " + device.getSystem());
                        System.out.println("    Current mix: " + device.getMix());
                        System.out.println("    Input channels: " + device.getChannelCount(false));
                        System.out.println("    Output channels: " + device.getChannelCount(true));
                        System.out.println("    Interleaved access: " + device.supportsAccess(true));
                        System.out.println("    Non-interleaved access: " + device.supportsAccess(false));
                    }
            }
        } catch (XtException e) {

            System.out.printf("Error: system %s, fault %s, cause %s, text %s, message: %s.\n",
                    XtException.getSystem(e.getError()),
                    XtException.getFault(e.getError()),
                    XtException.getCause(e.getError()),
                    XtException.getText(e.getError()),
                    e.toString());
        }
    }
}

File: src/cpp-sample/PrintDetailed.cpp
#include <xt/cpp/XtCpp.hpp>
#include <memory>
#include <cstdint>
#include <cstdlib>
#include <iostream>

static void OnError(const std::string& location, const std::string& message)
{ std::cout << location << ": " << message << std::endl; }

int PrintDetailedMain()
{
  auto audio = Xt::Audio::Init("Sample", nullptr, OnError);
  try 
  {
    Xt::Version version = Xt::Audio::GetVersion();
    std::cout << "Version: " << version.major << "." << version.minor << "\n";    
    Xt::System pro = Xt::Audio::SetupToSystem(Xt::Setup::ProAudio);
    std::cout << "Pro Audio: " << pro << " (" << (Xt::Audio::GetService(pro) != nullptr) << ")\n";
    Xt::System system = Xt::Audio::SetupToSystem(Xt::Setup::SystemAudio);
    std::cout << "System Audio: " << system << " (" << (Xt::Audio::GetService(system) != nullptr) << ")\n";
    Xt::System consumer = Xt::Audio::SetupToSystem(Xt::Setup::ConsumerAudio);
    std::cout << "Consumer Audio: " << consumer << " (" << (Xt::Audio::GetService(consumer) != nullptr) << ")\n";

    for(Xt::System s: Xt::Audio::GetSystems()) 
    {
      std::unique_ptr<Xt::Service> service = Xt::Audio::GetService(s);
      std::cout << "System " << s << ":\n";
      std::cout << "  Device count: " << service->GetDeviceCount() << "\n";
      std::cout << "  Capabilities: " << service->GetCapabilities() << "\n";
      std::unique_ptr<Xt::Device> defaultInput = service->OpenDefaultDevice(false);
      if(defaultInput) std::cout << "  Default input: " << *defaultInput << "\n";
      std::unique_ptr<Xt::Device> defaultOutput = service->OpenDefaultDevice(true);
      if(defaultOutput) std::cout << "  Default output: " << *defaultOutput << "\n";

      for(int32_t d = 0; d < service->GetDeviceCount(); d++)
      {
        std::unique_ptr<Xt::Device> device = service->OpenDevice(d);
        std::optional<Xt::Mix> mix = device->GetMix();
        std::cout << "  Device " << *device << ":" << "\n";
        std::cout << "    Input channels: " << device->GetChannelCount(false) << "\n";
        std::cout << "    Output channels: " << device->GetChannelCount(true) << "\n";
        std::cout << "    Interleaved access: " << device->SupportsAccess(true) << "\n";
        std::cout << "    Non-interleaved access: " << device->SupportsAccess(false) << "\n";
        if(mix) std::cout << "    Current mix: " << mix->rate << " " << mix->sample << "\n";
      }
    }
    return EXIT_SUCCESS;
  }
  catch(const Xt::Exception& e) 
  { 
    std::cout << Xt::Audio::GetErrorInfo(e.GetError()) << "\n"; 
    return EXIT_FAILURE;
  }
}